File: CSES/Dynamic Programming/1638_grids.cpp
// Problem Link: https://cses.fi/problemset/task/1638

//#include<bits/stdc++.h>

// #include </Users/prajwalm/Desktop/stdc++.h>
#include<bits/stdc++.h>
using namespace std;

vector<vector<char>> grids;
vector<vector<int>> cache;
int n;

int solve_i(){
    
    for(int i=n-1; i>=0; i--){
        for(int j=n-1; j>=0; j--){
 
            int mod = 1000000007;
 
            // base condition
            // at the bottom right corner
            if(i==n-1 && j==n-1){
                cache[i][j] =  1;
                continue;
            }
 
            // variables to keep track of obstacles and boundaries
            bool b_r = false; // boundary_right
            bool b_d = false; // boundary_down
            bool o_r = false; // obstacle_right
            bool o_d = false; // obstacle_down
 
            // Fill the variables defined above
            if(i==n-1){
                b_d = true;
            }
 
            if(j==n-1){
                b_r = true;
            }
    
            if(!b_d){
                if(grids[i+1][j] == '*'){
                    o_d = true;
                }
            }
 
            if(!b_r){
                if(grids[i][j+1] == '*'){
                    o_r = true;
                }
            }
 
    
            // compute the solution
            if(o_r && o_d){
            // if there is obstacle on both right and down move nowhere 
                cache[i][j] = 0;
            }
 
            else if (b_r){
                if (!o_d){
                // if on right boundary and there is no obstacle down
                // then move down
                    cache[i][j] = cache[i+1][j] % mod;
                }
            }
            else if (b_d){
                if (!o_r){
                // if on bottom boundary and there is no obstacle to the right
                // then move right
                    cache[i][j] = cache[i][j+1] % mod;
                }
            }
 
            else if (o_r){
                // if obstacle on right, move down
                cache[i][j] = cache[i+1][j] % mod;
            }
    
            else if (o_d){
                // if obstacle on down, move right
                cache[i][j] = cache[i][j+1] % mod;
            }
 
            else{
                // no obstacle on right or down
                cache[i][j] = ((cache[i+1][j] % mod)  + (cache[i][j+1] % mod)) % mod;
            }
        }
    }
 
    return cache[0][0];
}
 
int main(){

    cin>>n;
    
    for(int i=0; i<n; i++){
        vector<char> row;
        string k;
        cin>>k;
        for(char c: k){
            //cout<<c<<endl;
            row.push_back(c);
        }
        grids.push_back(row);
    }

    for(int i=0; i<n; i++){
        vector<int> row;
        for(int j=0; j<n; j++){
            row.push_back(0);
        }
        cache.push_back(row);
    }

    if(grids[0][0] == '*'){
        // special case when the start position itself is an obstacle
        cout<<0<<endl;
    }
    else{
        cout<<solve_i()<<endl;
    }

    return 0;

}

File: CSES/Dynamic Programming/1638_grids.py
# Problem Link: https://cses.fi/problemset/task/1638
 
# recursive solution
# solve(i, j) can be read out as move to cell defined by (i, j)
def solve_r(i, j):
 
    mod = 1000000007
 
    # return the solution if the problem is already solved
    if cache[i][j] != 0:
        return cache[i][j]
 
    # base condition
    # at the bottom right corner
    if (i==n-1) and (j==n-1):
        return 1
 
    # variables to keep track of obstacles and boundaries
    b_r = False # boundary_right
    b_d = False # boundary_down
    o_r = False # obstacle_right
    o_d = False # obstacle_down
 
    # Fill the variables defined above
    if i==n-1:
        b_d = True
 
    if j==n-1:
        b_r = True
    
    if not b_d:
        if grids[i+1][j] == '*':
            o_d = True
 
    if not b_r:
        if grids[i][j+1] == '*':
            o_r = True
 
    
    # compute the solution
    if o_r and o_d:
        # if there is obstacle on both right and down move nowhere 
        return cache[i][j]
 
    elif b_r:
        if not o_d:
            # if on right boundary and there is no obstacle down
            # then move down
            cache[i][j] = solve_r(i+1, j) % mod
 
    elif b_d:
        if not o_r:
            # if on bottom boundary and there is no obstacle to the right
            # then move right
            cache[i][j] = solve_r(i, j+1) % mod
 
    elif o_r:
        # if obstacle on right, move down
        cache[i][j] = solve_r(i+1, j) % mod
    
    elif o_d:
        # if obstacle on down, move right
        cache[i][j] = solve_r(i, j+1) % mod
 
    else:
        # no obstacle on right or down
        cache[i][j] = ((solve_r(i+1, j) % mod)  + (solve_r(i, j+1) % mod)) % mod
 
    return cache[i][j]
 
def solve_i():
 
    for i in range(n-1, -1, -1):
        for j in range(n-1, -1, -1):
 
            mod = 1000000007
 
            # base condition
            # at the bottom right corner
            if (i==n-1) and (j==n-1):
                cache[i][j] =  1
                continue
 
            # variables to keep track of obstacles and boundaries
            b_r = False # boundary_right
            b_d = False # boundary_down
            o_r = False # obstacle_right
            o_d = False # obstacle_down
 
            # Fill the variables defined above
            if i==n-1:
                b_d = True
 
            if j==n-1:
                b_r = True
    
            if not b_d:
                if grids[i+1][j] == '*':
                    o_d = True
 
            if not b_r:
                if grids[i][j+1] == '*':
                    o_r = True
 
    
            # compute the solution
            if o_r and o_d:
            # if there is obstacle on both right and down move nowhere 
                cache[i][j] = 0
 
            elif b_r:
                if not o_d:
                # if on right boundary and there is no obstacle down
                # then move down
                    cache[i][j] = cache[i+1][j] % mod
 
            elif b_d:
                if not o_r:
                # if on bottom boundary and there is no obstacle to the right
                # then move right
                    cache[i][j] = cache[i][j+1] % mod
 
            elif o_r:
                # if obstacle on right, move down
                cache[i][j] = cache[i+1][j] % mod
    
            elif o_d:
                # if obstacle on down, move right
                cache[i][j] = cache[i][j+1] % mod
 
            else:
                # no obstacle on right or down
                cache[i][j] = ((cache[i+1][j] % mod)  + (cache[i][j+1] % mod)) % mod
 
 
    return cache[0][0]
 
if __name__ == '__main__':

    from sys import stdin, stdout
    from time import perf_counter

    start = perf_counter()
    n = int(stdin.readline())
 
    grids = []
    cache = []
 
    # read the grid
    grids = [[j for j in stdin.readline()] for i in range(n)]

    # fill the cache for memoization
    cache = [ [0 for j in range(n)] for i in range(n)]
 
    if grids[0][0] == '*':
        # special case when the start position itself is an obstacle
        print(0)
    else:
        # print(solve_r(0,0))
        print(solve_i())
    
    stop = perf_counter()
    print(stop -start)
